// ngx-select-dropdown/esm2020/lib/ngx-select-dropdown.module.mjs
import { FormsModule } from "@angular/forms";
import { CommonModule } from "@angular/common";
import { NgModule } from "@angular/core";
import { NgxSelectDropdownComponent } from "./ngx-select-dropdown.component";
import { FilterByPipe } from "./pipes/filter-by.pipe";
import { LimitToPipe } from "./pipes/limit-to.pipe";
import * as i0 from "@angular/core";
export class SelectDropDownModule {
}
/** @nocollapse */ SelectDropDownModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.2.12", ngImport: i0, type: SelectDropDownModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
/** @nocollapse */ SelectDropDownModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "14.2.12", ngImport: i0, type: SelectDropDownModule, declarations: [NgxSelectDropdownComponent, FilterByPipe, LimitToPipe], imports: [CommonModule, FormsModule], exports: [NgxSelectDropdownComponent, FilterByPipe, LimitToPipe] });
/** @nocollapse */ SelectDropDownModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "14.2.12", ngImport: i0, type: SelectDropDownModule, imports: [CommonModule, FormsModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.2.12", ngImport: i0, type: SelectDropDownModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgxSelectDropdownComponent, FilterByPipe, LimitToPipe],
                    imports: [CommonModule, FormsModule],
                    exports: [NgxSelectDropdownComponent, FilterByPipe, LimitToPipe],
                    providers: [],
                }]
        }] });
