// ngx-select-dropdown/esm2020/lib/ngx-select-dropdown.service.mjs
import { EventEmitter, Injectable } from "@angular/core";
import * as i0 from "@angular/core";
export class SelectDropDownService {
    constructor() {
        this.openDropdownInstance = new EventEmitter();
        this.closeDropdownInstance = new EventEmitter();
        this.openInstances = [];
        // constructor
    }
    isOpen(instanceId) {
        return this.openInstances.indexOf(instanceId) > -1;
    }
    /**
     * @summary: Open a specific dropdown instance based on the instance ID.
     * @param instanceId: Instance id of the dropdown that must be opened.
     */
    openDropdown(instanceId) {
        this.openDropdownInstance.emit(instanceId);
    }
    /**
     * @summary: Close a specific dropdown instance based on the instance ID.
     * @param instanceId: Instance id of the dropdown that must be closed.
     */
    closeDropdown(instanceId) {
        this.closeDropdownInstance.emit(instanceId);
    }
}
/** @nocollapse */ SelectDropDownService.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.2.12", ngImport: i0, type: SelectDropDownService, deps: [], target: i0.ɵɵFactoryTarget.Injectable });
/** @nocollapse */ SelectDropDownService.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "14.2.12", ngImport: i0, type: SelectDropDownService, providedIn: "root" });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.2.12", ngImport: i0, type: SelectDropDownService, decorators: [{
            type: Injectable,
            args: [{
                    providedIn: "root",
                }]
        }], ctorParameters: function () { return []; } });
